---
title: "R Notebook"
output: html_notebook
---

The [R plugin](https://www.jetbrains.com/help/pycharm/r-plugin-support.html) for IntelliJ-based IDEs provides
handy capabilities to work with the [R Markdown](https://www.jetbrains.com/help/pycharm/r-markdown.html) files.
To [add](https://www.jetbrains.com/help/pycharm/r-markdown.html#add-code-chunk) a new R chunk,
position the caret at any line or the code chunk, then click "+".

The code chunk appears:
```{r, echo=F, print=F}
library(tidyverse)
library(data.table)
library(arrow)
library(ggpubr)
library(MatchIt)
```

Type any R code in the chunk, for example:
```{r}
# Gene data
gene_reference <- fread("/Users/cawarmerdam/Documents/projects/eQTLGen/public_data/Homo_sapiens.GRCh38.106.txt")
# Variant reference
variant_reference <- read_parquet("/Users/cawarmerdam/Documents/projects/eQTLGen/processed_data/variants/1000G-30x_index.parquet")
```

```{r}
eqtlgen <- fread("~/Downloads/eqtlgenphase2_chr21_pip09_caw_cl_noInDels-borzoi_RNA-blood_plot.tsv") %>%
  select(variant_index, beta, phenotype, diff_score) %>%
  inner_join(variant_reference, by = join_by(variant_index)) %>%
  inner_join(gene_reference, by = c("phenotype" = "gene_id")) %>% rename(gene_id = phenotype) %>%
  filter(as.character(chromosome) == seqid) %>%
  mutate(tss_distance = bp - start)

gtex <- fread("~/Downloads/2025-02-05_borzoi_eQTLs_noInDels_vs_borzoi_RNA-blood_chr21_only_plot.tsv") %>%
  mutate(gene_id = str_split_i(Gene, fixed("."), 1)) %>%
  select(variant_id, beta, gene_id, diff_score, tss_distance) %>%
  filter(!is.na(diff_score))

```

```{r}
res <- fread("~/Documents/projects/eQTLGen/fine_mapping/ExtractMetaAnalysisResults/carma_susie/combined_finemapping_output_5_runs.tsv")

borzoi <- fread("~/Downloads/unique_CS_variant_gene_combinations_annotated_noInDels-borzoi_RNA-blood.tsv.gz")
borzoi

res_borzoi <- res %>% filter(!is.na(SusieRss_CS), SusieRss_pip >= 0.9) %>% mutate(variant_id = paste0(phenotype, "_", variant_index)) %>% inner_join(borzoi)

ggplot(res_borzoi, aes(beta, logSED)) + geom_point(aes(colour = sample_size, size = abs(beta/standard_error)), shape = 16) + geom_smooth(method="lm") + stat_cor(method="spearman", cor.coef.name="rho") + stat_cor(method="pearson", label.y.npc=0.9, cor.coef.name="R") + facet_wrap(vars(fm_run_name))

eqtlgen <- res_borzoi %>%
  select(variant_index, beta, phenotype, diff_score = logSED, fm_run_name) %>%
  inner_join(variant_reference, by = join_by(variant_index)) %>%
  inner_join(gene_reference, by = c("phenotype" = "gene_id")) %>% rename(gene_id = phenotype) %>%
  filter(as.character(chromosome) == seqid) %>%
  mutate(tss_distance = bp - start)
```

```{r}
bound <- bind_rows(eqtlgen = eqtlgen, gtex = gtex %>% mutate(fm_run_name = "gtex"))

p <- bound %>% ggplot(aes(beta, diff_score)) + geom_point(shape = 16) + geom_smooth(method="lm") + stat_cor(method="spearman", cor.coef.name="rho") + stat_cor(method="pearson", label.y.npc=0.9, cor.coef.name="R") + facet_wrap(vars(fm_run_name))

plot(p)

# bound_gene_paired <- bound %>%
#   group_by(gene_id) %>% filter(n() == 2) %>% ungroup()
#
# p_paired <- bound_gene_paired %>% ggplot(aes(beta, diff_score)) + geom_point(shape = 16) + geom_smooth(method="lm") + stat_cor(method="spearman") + stat_cor(method="pearson", label.y.npc=0.9) + facet_grid(rows=vars(fm_run_name), cols=vars(dataset))

# plot(p_paired)

bound %>% ggplot(aes(fm_run_name, tss_distance)) + geom_violin() + geom_boxplot(width=0.1)

matched_df_list <- list()
for (fm_run_id in fm_run_ids) {
  matchit_input <- bound %>% filter(fm_run_name %in% c(fm_run_id, "gtex")) %>% mutate(gtex_lab=fm_run_name == "gtex")
  match_model <- matchit(gtex_lab ~ tss_distance, data = matchit_input, method = "nearest", ratio = 1)
  matched_df <- match.data(match_model)
  matched_df_list[[fm_run_id]] <- matched_df %>% filter(!gtex_lab)
}

matched_df_list[["gtex"]] <- bound %>% filter(fm_run_name == "gtex")
matched_df <- bind_rows(matched_df_list)


matched_df %>% ggplot(aes(fm_run_name, tss_distance)) + geom_violin() + geom_boxplot(width=0.1)


tss_mtched <- matched_df %>% ggplot(aes(beta, diff_score)) + geom_point(shape = 16) + geom_smooth(method="lm") + stat_cor(method="spearman", cor.coef.name="rho") + stat_cor(method="pearson", label.y.npc=0.9, cor.coef.name="R") + facet_wrap(vars(fm_run_name))

plot(tss_mtched)
```